bee-signing/src/ternary/seed.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! Ternary seed to derive private keys, public keys and signatures from.
+
+use bee_common_derive::{SecretDebug, SecretDisplay, SecretDrop};
+use bee_crypto::ternary::{
+    sponge::{Kerl, Sponge},
+    HASH_LENGTH,
+};
+use bee_ternary::{Btrit, T1B1Buf, Trit, TritBuf, Trits, TryteBuf, T1B1};
+
+use rand::Rng;
+use thiserror::Error;
+use zeroize::Zeroize;
+
+use std::str::FromStr;
+
+/// Errors occuring when handling a `Seed`.
+#[derive(Debug, Error, PartialEq)]
+pub enum Error {
+    /// Invalid seed length.
+    #[error("Invalid seed length.")]
+    InvalidLength(usize),
+    /// Invalid seed trytes.
+    #[error("Invalid seed trytes.")]
+    InvalidTrytes,
+    /// Failed sponge operation.
+    #[error("Failed sponge operation.")]
+    FailedSpongeOperation,
+}
+
+/// Ternary `Kerl`-based `Seed` to derive private keys, public keys and signatures from.
+#[derive(SecretDebug, SecretDisplay, SecretDrop)]
+pub struct Seed(TritBuf<T1B1Buf>);
+
+impl Zeroize for Seed {
+    fn zeroize(&mut self) {
+        // This unsafe is fine since we only reset the whole buffer with zeros, there is no alignement issues.
+        unsafe { self.0.as_i8_slice_mut().zeroize() }
+    }
+}
+
+impl Seed {
+    /// Creates a new random `Seed`.
+    pub fn rand() -> Self {
+        // `ThreadRng` implements `CryptoRng` so it is safe to use in cryptographic contexts.
+        // https://rust-random.github.io/rand/rand/trait.CryptoRng.html
+        let mut rng = rand::thread_rng();
+        let trits = [Btrit::NegOne, Btrit::Zero, Btrit::PlusOne];
+        let mut seed = [Btrit::Zero; HASH_LENGTH];
+
+        for trit in seed.iter_mut() {
+            *trit = trits[rng.gen_range(0, trits.len())];
+        }
```

Comment:
I propose to use rand's `sample` API instead. It much cleaner IMHO, don't know if it also has performance benefits. I just find `gen_range` in a loop not so elegant here.

Reply:
Can you give me an example with `sample` ? Not sure to see how to use it here.

Reply:
okay, os you basically want to uniformly random select between the 3 possible trit values.
https://docs.rs/rand/0.7.3/rand/distributions/uniform/struct.Uniform.html#method.sample

Reply:
so from reading the docs it seems to actually be faster and also less biased.

Reply:
Alright, done.